### PR TITLE
Clean takes a flag that will add a --force flag to clean.

### PR DESCRIPTION
**Pull request by roguishmountain**
See bug report https://issues.jenkins-ci.org/browse/JENKINS-26660 for more details

### CODE REVIEW

**Comment by roguishmountain** [comment]
Depends on https://github.com/jenkinsci/git-plugin/pull/449

**Comment by roguishmountain** [comment]
rebuild

**Comment by roguishmountain** [comment]
Hi @MarkEWaite, happy Thanksgiving, we’d really like to get this PR and the dependent PR in git-plugin merged. Let me know if anything needs to be fixed or changed.

**Comment by roguishmountain** [comment]
Hi @MarkEWaite, just wanted to check in again. Do you know when you will be able to look over this PR and PR in the git-plugin?

**Comment by MarkEWaite** [comment]
I am confident that I will not evaluate it for at least a month.  I'm working now on PR225 to resolve the renamed submodule regression that was introduced in git plugin 3.0.  Once tests are written to resolve that regression, I'll next focus my efforts on resolving the authentication regression(s) that were introduced in git plugin 3.0.  Once those two are completed, I'll review other pull requests to see which should be reviewed next.

I'd love to have you enlist others to help test this pull request and its matching pull request for the git plugin.  There are a few different Jira reports requesting this feature, so you might comment on those bug reports to see if you can persuade the submitters of those bugs to help confirm the usefulness and compatibility of this pull request.

I'm more readily persuaded when multiple users report that they are successfully running the pull request

**Comment by leandro-lucarella-sociomantic** [comment]
This would fix https://issues.jenkins-ci.org/browse/JENKINS-45705 for me.

File: src/main/java/org/jenkinsci/plugins/gitclient/CliGitAPIImpl.java [hunk 1]
```diff
@@ -682,6 +682,20 @@ public void clean(boolean cleanSubmodule) throws GitException, InterruptedExcept
...
+     *
+     * @throws hudson.plugins.git.GitException if underlying git operation fails.
+     * @throws java.lang.InterruptedException if interrupted.
+     */
+    public void clean() throws GitException, InterruptedException {
+        reset(true);
+        String cmd = "-fdx";
+
+        launchCommand("clean", cmd);
+    }
```
Comment (leandro-lucarella-sociomantic) [inline src/main/java/org/jenkinsci/plugins/gitclient/CliGitAPIImpl.java]: I want to first clarify I'm not a Java programmer, so what I say may not make sense or not be a common practice in Java, but can't you just set a default parameter `public void clean(boolean cleanSubmodule = false)`?

**Comment by MarkEWaite** [comment]
@leandro-lucarella-sociomantic you can test drive the plugin by downloading it from the "all checks have passed" [details link](https://ci.jenkins.io/job/Plugins/job/git-client-plugin/job/PR-222/110/display/redirect)  

That takes you to a page which has an "Artifacts" link in the top right corner.  When you click that link, one of the items to download is [git-client.hpi](https://ci.jenkins.io/job/Plugins/job/git-client-plugin/job/PR-222/110/artifact/target/git-client.hpi).  That is the tip of the current master branch with this pull request merged onto it.

**Comment by leandro-lucarella-sociomantic** [comment]
Thanks! I guess I will need to install a matching github-plugin in https://github.com/jenkinsci/git-plugin/pull/449 for this right? But that PR is not building so I can't download it.

**Comment by leandro-lucarella-sociomantic** [comment]
ping @MarkEWaite

**Comment by MarkEWaite** [comment]
@leandro-lucarella-sociomantic I won't be able to assist with this until after the end of the working day today, or possibly tomorrow.  Too many things to do for work today.

**Comment by leandro-lucarella-sociomantic** [comment]
Just for the records, we had deployed the plugin with this fix for maybe like a month or even more. Sadly I couldn't get the people affected by this to test it in this time, but on the other hand, nothing else broke, so I guess that is in a way good news :)

**Pull request merged**